### PR TITLE
Fixes autotest runs on mac osx

### PR DESCRIPTION
**Pull request by bugobliterator**
Depends on https://github.com/ArduPilot/ardupilot/pull/14362

### CODE REVIEW

File: Tools/autotest/autotest.py [hunk 1]
```diff
@@ -650,8 +650,8 @@ def list_subtests(*args, **kwargs):
 if __name__ == "__main__":
     ''' main program '''
     os.environ['PYTHONUNBUFFERED'] = '1'
-
-    os.putenv('TMPDIR', util.reltopdir('tmp'))
+    if sys.platform == "TMPDIR":
```
Comment (khancyr) [inline Tools/autotest/autotest.py:653]: if sys.platform != 'darwin' ?

Comment (peterbarker) [reply Tools/autotest/autotest.py:653]: Or, you know, kill the delta since it's apparently not required? :-)

File: libraries/AP_Scripting/generator/src/main.c [hunk 2]
```diff
@@ -673,7 +673,7 @@ void handle_operator(struct userdata *data) {
     error(ERROR_USERDATA, "Needed a symbol for the operator");
   }
 
-  enum operator_type operation;
+  enum operator_type operation = OP_ADD;
```
Comment (peterbarker) [inline libraries/AP_Scripting/generator/src/main.c:676]: Could this be done better with a NORETURN or whatever attribute on "error"?

File: Tools/autotest/pysim/util.py [hunk 3]
```diff
@@ -309,7 +325,9 @@ def start_SITL(binary,
         f.write("settings set target.process.stop-on-exec false\n")
         f.write("process launch\n")
         f.close()
-        if os.environ.get('DISPLAY'):
+        if sys.platform == "darwin" and os.getenv('DISPLAY'):
+            cmd.extend(['lldb', '-s', '/tmp/x.lldb', '--'])
```
Comment (peterbarker) [inline Tools/autotest/pysim/util.py:329]: Why aren't we starting this in a new graphical terminal?

File: Tools/autotest/pysim/util.py [hunk 4]
```diff
@@ -291,7 +305,9 @@ def start_SITL(binary,
             f.write("disable\n")
         f.write("r\n")
         f.close()
-        if os.environ.get('DISPLAY'):
+        if sys.platform == "darwin" and os.getenv('DISPLAY'):
+            cmd.extend(['gdb', '-x', '/tmp/x.gdb', '--args'])
```
Comment (peterbarker) [inline Tools/autotest/pysim/util.py:309]: Why aren't we starting this in a new graphical terminal?

File: Tools/autotest/pysim/util.py [hunk 5]
```diff
@@ -230,8 +237,15 @@ def valgrind_log_filepath(binary, model):
 
 
 def kill_screen_gdb():
-    cmd = ["screen", "-X", "-S", "ardupilot-gdb", "quit"]
-    subprocess.Popen(cmd)
+    global windowID
+    if sys.platform == "darwin":
```
Comment (peterbarker) [inline Tools/autotest/pysim/util.py:241]: I'm not really sure why we're closing windows here.  We use screen when we *don't* have a windowing system

File: libraries/AP_Math/matrix_alg.cpp [hunk 6]
```diff
@@ -285,7 +285,7 @@ bool inverse4x4(float m[],float invOut[])
     uint8_t i;
 
 #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
-    int old = fedisableexcept(FE_OVERFLOW);
+    int old = fedisableexcept(FE_INEXACT | FE_OVERFLOW);
```
Comment (peterbarker) [inline libraries/AP_Math/matrix_alg.cpp:288]: Why is this required?  Why doesn't this fail in normal SITL?

File: Tools/autotest/autotest.py (hunk 1 above)
Comment (bugobliterator) [reply Tools/autotest/autotest.py:653]: yeap I will remove this line entirely

File: libraries/AP_Scripting/generator/src/main.c (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_Scripting/generator/src/main.c:676]: This change already went in apparently.

File: libraries/AP_Math/matrix_alg.cpp (hunk 6 above)
Comment (bugobliterator) [reply libraries/AP_Math/matrix_alg.cpp:288]: Yeah, I can only reproduce this inside mac, and it happens every time. Instructions used by mac os for floating point calculations are different.

File: Tools/autotest/pysim/util.py (hunk 5 above)
Comment (bugobliterator) [reply Tools/autotest/pysim/util.py:241]: For mac os I made a change to start debugger on a separate terminal.

File: Tools/autotest/pysim/util.py (hunk 4 above)
Comment (bugobliterator) [reply Tools/autotest/pysim/util.py:309]: we do that down below https://github.com/ArduPilot/ardupilot/pull/14363/files/af2fbe4047435c3bdb3bcc987fd3ce92cb18b865#diff-d2de282cca7753bde1b8719902d06dcbR355

File: Tools/autotest/pysim/util.py (hunk 3 above)
Comment (bugobliterator) [reply Tools/autotest/pysim/util.py:329]: we do that down below https://github.com/ArduPilot/ardupilot/pull/14363/files/af2fbe4047435c3bdb3bcc987fd3ce92cb18b865#diff-d2de282cca7753bde1b8719902d06dcbR355

File: Tools/autotest/pysim/util.py (hunk 5 above)
Comment (bugobliterator) [reply Tools/autotest/pysim/util.py:241]: right I should check for DISPLAY here.

**Comment by peterbarker** [comment]
On Wed, 13 May 2020, Siddharth Bharat Purohit wrote:

> For mac os I made a change to start debugger on a separate terminal.

The point here was that separate graphical elements should never be in 
play at the same time that you're using GNU screen.  The idea is that if 
you don't have a graphical environment (ssh'd in with no forwarding, for 
example), you can still work with gdb if you have screen available.

**Comment by bugobliterator** [comment]
> On Wed, 13 May 2020, Siddharth Bharat Purohit wrote: For mac os I made a change to start debugger on a separate terminal.
> The point here was that separate graphical elements should never be in play at the same time that you're using GNU screen. The idea is that if you don't have a graphical environment (ssh'd in with no forwarding, for example), you can still work with gdb if you have screen available.

Yeah, it will still fall through to screen if there's no graphical interface even in mac.

**Comment by bugobliterator** [comment]
@peterbarker Is this good to go in?

File: Tools/autotest/autotest.py [hunk 7]
```diff
@@ -651,8 +651,6 @@ def list_subtests(*args, **kwargs):
     ''' main program '''
     os.environ['PYTHONUNBUFFERED'] = '1'
 
-    os.putenv('TMPDIR', util.reltopdir('tmp'))
```
Comment (peterbarker) [inline Tools/autotest/autotest.py:654]: Why does this break MacOSX stuff?

This wouldn't have been put here for fun, I'm sure someone needed this.  For example, on the autotest server we use a specific partition for building for speed, using the default `TMPDIR` may be slow.

File: Tools/autotest/pysim/util.py [hunk 8]
```diff
@@ -173,6 +177,9 @@ def pexpect_close(p):
     global close_list
 
     ex = None
+    if p is None:
```
Comment (peterbarker) [inline Tools/autotest/pysim/util.py:180]: Who is calling this with`p` as None?

File: libraries/AP_Math/matrix_alg.cpp [hunk 9]
```diff
@@ -405,12 +405,6 @@ bool inverse4x4(float m[],float invOut[])
 
     det = m[0] * inv[0] + m[1] * inv[4] + m[2] * inv[8] + m[3] * inv[12];
 
-#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Comment (peterbarker) [inline libraries/AP_Math/matrix_alg.cpp:408]: Do we have to move the re-enabling of `FE_OVERFLOW` down as well to get things working on MacOSX?

File: libraries/AP_Math/matrix_alg.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_Math/matrix_alg.cpp:288]: OK, could you augment the commit message, please?

File: Tools/autotest/autotest.py (hunk 7 above)
Comment (bugobliterator) [reply Tools/autotest/autotest.py:654]: @peterbarker the ar on mac doesn't like setting of TMPDIR. mac os has its own system of tmpdir happening. I can put this back in for other systems. I removed it because you said it wasn't necessary.

File: Tools/autotest/pysim/util.py (hunk 8 above)
Comment (bugobliterator) [reply Tools/autotest/pysim/util.py:180]: On mac with display, the gdb lldb is launched using `subprocess.Popen` instead of pexpect.

File: libraries/AP_Math/matrix_alg.cpp (hunk 9 above)
Comment (bugobliterator) [reply libraries/AP_Math/matrix_alg.cpp:408]: Yes, there was another failure detected below.

File: Tools/autotest/autotest.py (hunk 7 above)
Comment (bugobliterator) [reply Tools/autotest/autotest.py:654]: I have put this back in for systems other than macos

**Comment by peterbarker** [comment]
On Sat, 16 May 2020, Siddharth Bharat Purohit wrote:

> On mac with display, the gdb lldb is launched using subprocess.Popen instead of pexpect.

Do the windows still get closed OK on exit?

**Comment by bugobliterator** [comment]
> On Sat, 16 May 2020, Siddharth Bharat Purohit wrote: On mac with display, the gdb lldb is launched using subprocess.Popen instead of pexpect.
> Do the windows still get closed OK on exit?
@peterbarker Yes, the process for macos is bit different, but follows the policy followed in normal SITL run.

**Comment by peterbarker** [comment]
Merged - thanks!
